### ISSUE TITLE
Angular3 `pub build` causes "Constant expressions can't refer to type variables."

### ISSUE THREAD

**Issue by zoechi** [dart-lang/sdk#28551]
```
[Error from Dart2JS on test_project|web/index.dart]:
packages/observable/src/collections/observable_set.dart:159:38:
Constant expressions can't refer to type variables.
Try removing the type variable or replacing it with a concrete type.
  Iterator<E> get iterator => const <E>[].iterator;
```
I could successfully build other more complex but otherwise similar projects.

> Dart VM version: 1.22.0-edge.0e00b3e5c83e4173d442bf287eb99d1dd12cb09b (Wed Jan 25 15:35:49 2017) on "linux_x64"

Edit (floitschG): Link to `observable` bug:
https://github.com/dart-lang/observable/issues/31

**Comment by zoechi** [comment]
ups, built wrong project.

**Comment by zoechi** [comment]
https://github.com/dart-lang/observable/blob/master/lib/src/collections/observable_set.dart#L159
Seems not to cause an error in Analyzer or VM, only `pub build`.

**Comment by a-siva** [comment]
Can you add some reproduction instructions here

**Comment by zoechi** [comment]
I don't use `observable` directly. It's somehow used by Angular. 
Should this code work or is it invalid as the error message says?

**Comment by a-siva** [comment]
The code is invalid, see a small test - 
```
class A<E> {
  Iterator<E> get iterator => const <E>[].iterator;
}
main() {
  A<int> a = new A<int>();
  Iterator<int> i1 = a.iterator;
}
```
This produces the error:
```
error: line 2 pos 37: the type argument of a constant list literal cannot include a type variable
  Iterator<E> get iterator => const <E>[].iterator;
```

**Comment by zoechi** [comment]
I don't get any complaints for the code linked above from the Analyzer in WebStorm when I open the `observable` project.

```
Dart VM version: 1.23.0-edge.b83cabe79c3703346d2134e80f7b63a9c1c18eff (Fri Feb  3 03:42:54 2017) on "linux_x64"
```

**Comment by a-siva** [comment]
Do you use the --strong option when analyzing WebStorm code?

**Comment by eernstg** [comment]
A const expression `e` must denote a single value such that the value can be created at compile time and used as the value of all const expressions denoting the same value (which certainly includes `e` itself, plus possibly some textually identical expressions elsewhere, plus possibly some textually different expressions with the same value). If `E` denotes two different values at two different occasions at run time, `const <E>[].iterator` couldn't possibly be the same object. So it will have to be handled differently.

One thing you can always consider is whether you could use a subtype: `const <Null>[].iterator` seems to work here. In general, `Null` as a type argument may help you out when you want to provide a "subtype of everything". Otherwise you might just change the whole expression to not be const.

The reason why the analyzer doesn't complain is probably that the DDC has been experimenting with a slightly different notion of `const` values, where canonicalization happens at run time and hence run-time values can be taken into account. But we won't promise you that this will be the future treatment of const so considering const as strictly "one expression, one thing" is the safe bet.

**Comment by floitschG** [comment]
TL;DR: the observable package should change the `const <E>[].iterator` to `const <Null>[].iterator`.

This is a transitional problem: strong mode (and DDC) currently allow constants to refer to generic types as if they were constant. (DDC dynamically canonicalizes).
Dart 1 (and the current versions of dart2js and the VM) don't have access to `E` since it's not known at compile time.

The current behavior of strong mode is to canonicalize dynamically (at least for DDC), but we are working on changing this to infer bottom (that is `Null`, or later `Nothing`) for the type if a `const` expression uses a generic type.

Instead of relying on that inference change, the observable package should simply change the generic type inside the constant expression and replace the use of `E` with `Null`. I will follow up with them.

I just created a bug to track the transition to `Null` and `Nothing`: https://github.com/dart-lang/sdk/issues/28685

**Comment by floitschG** [comment]
And thanks!

**Comment by floitschG** [comment]
Correcting myself: we probably never want to infer `Null` for a const expression that explicitly refers to `E`. However, without type (relying on the type-inference) we want to find `Null`.

**Comment by a-siva** [comment]
Is there a bug to track the necessary changes needed to the observable package ?

**Comment by zoechi** [comment]
> Do you use the --strong option when analyzing WebStorm code?

In my own projects I use `analysis_options.yaml` to enable strong mode, when I checked out the `observable` package, I just used it as is.
I'm just on the phone right now. I'll check again with strong mode enabled when I'm back in a few days.

**Comment by floitschG** [comment]
Observable package bug was already added to this issue (see my edit), and it was already fixed.

**Comment by zoechi** [comment]
@a-siva The observable package contains an `.analysis_options` file with 

```yaml
analyzer:
  strong-mode: true
```
but there is no hint for the type parameter.

**Comment by ziedHamdi** [comment]
I have a similar issue on ubuntu dart 1.24.3

[Error from Dart2JS on share_place|web/volatile_tests.dart]:
web/packages/observable/src/collections/observable_list.dart:65:18:
Constant expressions can't refer to type variables.
Try removing the type variable or replacing it with a concrete type.
    return const ListDiffer<T>().diff(previous, current);
                 ^^^^^^^^^^^^^
[Warning from Dart2JS on share_place|web/volatile_tests.dart]:
4 hint(s) suppressed in package:angular2.
[Warning from Dart2JS on share_place|web/volatile_tests.dart]:
2 hint(s) suppressed in package:collection.
[Info from Dart2JS]:
Took 0:00:07.207256 to compile share_place|web/volatile_tests.dart.
Build failed.


volatile_tests.dart is pretty simple:
import 'package:angular2/core.dart';
import 'package:angular2/platform/browser.dart';
import 'package:share_place/tests/tests_comp.dart';
import 'package:share_place/event_bus.dart';
import 'package:share_place/environment.dart';
import 'package:share_place/browser_data_service.dart';
import 'package:http/http.dart';
/**
 * See those subjects :
 *  - https://angular.io/docs/ts/latest/guide/router.html#!#base-href
 *
 */

void main() {
  bootstrap(
      TestsComp,
      [provide(Client, useClass: BrowserDataService), Environment, EventBus]
  );
}
